packages/gatsby/index.d.ts

```diff
@@ -387,8 +387,8 @@ export interface GatsbyNode {
   resolvableExtensions?(
     args: ResolvableExtensionsArgs,
     options: PluginOptions,
-    callback: PluginCallback
-  ): any[] | Promise<any[]>
+    callback?: PluginCallback<string[]>
+  ): string[]
```

Comment:
Should this be
```suggestion
  ): string[] | Promise<string[]>
```
? Every gatsby-node API hook can be either sync or async. 

Another thought - API hooks should either return values or use callback - with this function definition users/plugins would be forced to return array even if they would want to use callback? (not that it's working fine right now, it certainly is broken too)

Reply:
Some APIs have 3 overloads, e.g. `sourceNodes` is declared as 

```
  sourceNodes?(args: SourceNodesArgs, options: PluginOptions): void
  sourceNodes?(args: SourceNodesArgs, options: PluginOptions): Promise<void>
  sourceNodes?(
    args: SourceNodesArgs,
    options: PluginOptions,
    callback: PluginCallback<void>
  ): void
```

While others like `resolvableExtensions` have only one definition... I can add the Promise overload, but why 3 overloads instead of two? can't we write this (just making `callback` optional):

```
  sourceNodes?(args: SourceNodesArgs, options: PluginOptions): Promise<void>
  sourceNodes?(
    args: SourceNodesArgs,
    options: PluginOptions,
    callback?: PluginCallback<void>
  ): void
```

Reply:
I don't have context why there are 3 overloads and not just 2, but it seems sensible to be able to just have 2 with optional `callback` (at least for API hooks that are not meant to return anything)

Reply:
so in this case, maybe we should write this?
```
  resolvableExtensions?(
    args: ResolvableExtensionsArgs,
    options: PluginOptions,
  ): Promise<string[]>
  resolvableExtensions?(
    args: ResolvableExtensionsArgs,
    options: PluginOptions,
    callback?: PluginCallback<string[]>
  ): void
```

I mean according to the general API hooks documentation we should not even return a `string[]`, it should be either through the callback or through a Promise, right? BUT, [in this special case, the documentation says](https://www.gatsbyjs.com/docs/node-apis/#resolvableExtensions) 
> Return value: `string[]`

Which is why I originally typed the hook with just `string[]` as a return value.

Reply:
Right, documentation needs adjustment too :(

It's surprisingly hard to write those docs in readable way given that function can be either sync or async AND that `callback` option is also there :/

Reply:
So let's be clear: APIs have 3 signatures?
- sync, using `return value`
- sync, using `callback(value)`
- async, using `return Promise(value)` (or `async` equivalent)

Because the first signature availability is not obvious to me reading the docs.

Also, do all APIs have those 3 possibilities or do we have some APIs having only sync signatures?

Reply:
I guess you can say 4, because callback can be either sync or async (if `callback` is used we ignore returned value - https://github.com/gatsbyjs/gatsby/blob/b6cde3bd5524c7a820a31a4d28fa6f34058ce1bf/packages/gatsby/src/utils/api-runner-node.js#L327-L357) - but that's just for `gatsby-node` API hooks. Also the comment with "expect a Promise to be returned" is not quite accurate - it doesn't need to be promise - it can be just sync value (`await` can handle it)

I think the split should be first depending of `callback` is used or not and if it's not used then we expect either sync or async function returning specified type (which is `void` in most cases)

I've been looking more through those types and I see now that some that shouldn't even have `callback` variant do have those right now (the `gatsby-ssr` and `gatsby-browser` actually don't support those) - so my comment on `wrapPageElement` is actually wrong - there shouldn't be `callback` option at all there :/